cmd/dataset_tokenizer/dataset_tokenizer.go

```diff
@@ -502,17 +507,79 @@ func WriteContexts(outPath string, nextContext ContextsIterator,
 		}
 	}()
 
-	for {
-		context, more := <-contexts
-		if !more {
-			break
+	endpos := 0
+	var buf []byte
+	var contextSize int
+	var target int64
+
+	if sortSpec == "shuffle" {
+		fmt.Printf("Shuffling!\n")
+		for {
+			context, more := <-contexts
+			if !more {
+				break
+			}
+			binContext := context.ToBin()
+			if endpos == 0 {
+				contextSize = len(*binContext)
+				buf = make([]byte, contextSize)
+
+			}
+			if endpos >= 0 {
+				if endpos == 0 {
+					target = 0
+				} else if endpos == 1 {
+					target = int64(contextSize)
+				} else {
+					target = int64(rand.Intn(endpos-1) * contextSize)
+				}
+
+				_, err = outFile.Seek(target, 0)
+				if err != nil {
+					log.Fatal(err)
+				}
+				if endpos > 1 {
+					_, err = outFile.Read(buf)
+					if err != nil {
+						log.Fatal(err)
+					}
+				}
+				_, err = outFile.Seek(target, 0)
+				if err != nil {
+					log.Fatal(err)
+				}
+				if _, writeErr := outFile.Write(*binContext); writeErr != nil {
+					return totalTokens, writeErr
+				}
+
+				_, err = outFile.Seek(int64(endpos*contextSize), 0)
+				if err != nil {
+					log.Fatal(err)
+				}
+				if endpos > 1 {
+					if _, writeErr := outFile.Write(buf[0:contextSize]); writeErr != nil {
+						return totalTokens, writeErr
+					}
+				}
+				totalTokens += len(context)
+			}
+			endpos += 1
+
 		}
-		binContext := context.ToBin()
-		if _, writeErr := outFile.Write(*binContext); writeErr != nil {
-			return totalTokens, writeErr
+	} else {
+		for {
+			context, more := <-contexts
```

Comment:
We have some repeated code here. Could this instead be one loop, and we check for a `shuffle` boolean value?